ksm.go

```diff
@@ -0,0 +1,497 @@
+		return "", err
+	}
+
+	pagesToScan := nPages / s.pagesPerScanFactor
+
+	return fmt.Sprintf("%v", pagesToScan), nil
+}
+
+type ksmMode string
+
+const (
+	ksmInitial    ksmMode = "initial"
+	ksmOff        ksmMode = "off"
+	ksmSlow       ksmMode = "slow"
+	ksmStandard   ksmMode = "standard"
+	ksmAggressive ksmMode = "aggressive"
+	ksmAuto       ksmMode = "auto"
+)
+
+var ksmSettings = map[ksmMode]ksmSetting{
+	ksmOff:        {false, 1000, 500}, // Turn KSM off
+	ksmSlow:       {true, 500, 100},   // Every 100ms, we scan 1 page for every 500 pages available in the system
+	ksmStandard:   {true, 100, 10},    // Every 10ms, we scan 1 page for every 100 pages available in the system
+	ksmAggressive: {true, 10, 1},      // Every ms, we scan 1 page for every 10 pages available in the system
+}
+
+func (k ksmMode) String() string {
+	switch k {
+	case ksmOff:
+		return "off"
+	case ksmInitial:
+		return "initial"
+	case ksmAuto:
+		return "auto"
+	}
+
+	return ""
+}
+
+func (k *ksmMode) Set(value string) error {
+	for _, r := range strings.Split(value, ",") {
+		if r == "off" {
+			*k = ksmOff
+			return nil
+		} else if r == "initial" {
+			*k = ksmInitial
+			return nil
+		} else if r == "auto" {
+			*k = ksmAuto
+			return nil
+		}
+
+		return errors.New("Unsupported KSM knob")
+	}
+
+	return nil
+}
+
+var defaultKSMRoot = "/sys/kernel/mm/ksm/"
+var memInfo = "/proc/meminfo"
+
+const (
+	ksmRunFile       = "run"
+	ksmPagesToScan   = "pages_to_scan"
+	ksmSleepMillisec = "sleep_millisecs"
+	ksmStart         = "1"
+	ksmStop          = "0"
+	defaultKSMMode   = ksmAuto
+)
+
+type sysfsAttribute struct {
+	path string
+	file *os.File
+}
+
+func (attr *sysfsAttribute) open() error {
+	file, err := os.OpenFile(attr.path, os.O_RDWR|syscall.O_NONBLOCK, 0660)
+	attr.file = file
+	return err
+}
+
+func (attr *sysfsAttribute) close() error {
+	err := attr.file.Close()
+	attr.file = nil
+	return err
+}
+
+func (attr *sysfsAttribute) read() (string, error) {
+	_, err := attr.file.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return "", err
+	}
+
+	data, err := ioutil.ReadAll(attr.file)
+	if err != nil {
+		return "", err
+	}
+
+	return string(data), nil
+}
+
+func (attr *sysfsAttribute) write(value string) error {
+	_, err := attr.file.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return err
+	}
+
+	err = attr.file.Truncate(0)
+	if err != nil {
+		return err
+	}
+
+	_, err = attr.file.WriteString(value)
+
+	return err
+}
+
+type ksm struct {
+	root          string
+	run           sysfsAttribute
+	pagesToScan   sysfsAttribute
+	sleepInterval sysfsAttribute
+	initialized   bool
+
+	initialPagesToScan   string
+	initialSleepInterval string
+	initialKSMRun        string
+
+	currentKnob ksmMode
+	throttling  bool
+	kickChannel chan bool
+
+	sync.Mutex
+}
+
+func (k *ksm) isAvailable() error {
+	info, err := os.Stat(k.root)
+	if err != nil || !info.IsDir() {
+		return fmt.Errorf("%s is not available", k.root)
+	}
+
+	return nil
+}
+
+func (k *ksm) init(root string) error {
```

Comment:
OOI, is there a reason this couldn't have been done in a `func NewKSM(root string) (ksm, error)` function?